firmware/common/untar.hpp

```diff
@@ -0,0 +1,147 @@
+            i *= 8;
+            i += *p - '0';
+            ++p;
+            --n;
+        }
+        return i;
+    }
+
+    static bool is_end_of_archive(const char* p) {
+        for (int n = 511; n >= 0; --n)
+            if (p[n] != '\0') return false;
+        return true;
+    }
+
+    static bool create_dir(char* pathname) {
+        char* p;
+        std::filesystem::filesystem_error r;
+
+        /* Strip trailing '/' */
+        if (pathname[strlen(pathname) - 1] == '/')
+            pathname[strlen(pathname) - 1] = '\0';
+
+        /* Try creating the directory. */
+        std::string dirnameStr = u'/' + pathname;
+        std::filesystem::path dirname = dirnameStr;
+
+        r = make_new_directory(dirname);
+
+        if (!r.ok()) {
+            /* On failure, try creating parent directory. */
+            p = strrchr(pathname, '/');
+            if (p != NULL) {
+                *p = '\0';
+                create_dir(pathname);
+                *p = '/';
+                r = make_new_directory(dirname);
+            }
+        }
+        return (r.ok());
+    }
+
+    static bool verify_checksum(const char* p) {
+        int n, u = 0;
+        for (n = 0; n < 512; ++n) {
+            if (n < 148 || n > 155)
+                /* Standard tar checksum adds unsigned bytes. */
+                u += ((unsigned char*)p)[n];
+            else
+                u += 0x20;
+        }
+        return (u == parseoct(p + 148, 8));
+    }
+
+    static std::filesystem::path untar_int(File* a, std::function<void(const std::string)> cb = NULL) {
+        char buff[512];
+        UINT bytes_read;
+        std::string binfile = "";
+        std::string fn = "";
+        int filesize;
+        for (;;) {
+            auto readres = a->read(buff, 512);
+            if (!readres.is_ok()) return "";
+            bytes_read = readres.value();
+            if (bytes_read < 512) {
+                return "";
+            }
+            if (is_end_of_archive(buff)) {
+                return binfile;
+            }
+            if (!verify_checksum(buff)) {
+                return "";
+            }
+            filesize = parseoct(buff + 124, 12);
+            switch (buff[156]) {
+                case '1':
+                    break;
+                case '2':
+                    break;
+                case '3':
+                    break;
+                case '4':
+                    break;
+                case '5':
+                    create_dir(buff);
+                    filesize = 0;
+                    break;
+                case '6':
+                    break;
+                default:
+                    fn = buff;
+                    if (fn.length() > 5 && fn.substr(fn.length() - 4) == ".bin") {
+                        binfile = fn;
+                    }
+                    if (cb != NULL) cb(fn);
+                    break;
+            }
+            File f;
+            if (filesize > 0) {
+                delete_file(fn);
```

Comment:
Yeah, I was thinking about it, but if we would generate file names based on releases, we don't really have to worry about it. Also if I test or develop something, I really don't want to get a tons of fw files on the sd, but i would like to have them with a proper name.
Like portapack-h1_h2-mayhem_1.9.1.bin
or portapack-h1_h2-mayhem_daily_20240103.bin

Instead of portapack-h1_h2-mayhem-57.bin
I know we could use both, but in the end we would have tons of files like portapack-h1_h2-mayhem_1.9.1-54.bin, and those would be identical.

Also when using the 11th dev build I don't want to rollback, so overwriting it is better than the need of deleting them regularly.

But If you say It should be renamed, I'll implement it, just sharing my opinion.

Reply:
Yeah good point, not to flood the card with dev builds 👍